### PR TITLE
make session.unsubscribe command atomic

### LINKED ISSUE #482: `session.unsubscribe` should be atomic
The command `session.unsubscribe` should be atomic. If operation failed to some reason (not subscribed, browsing context not found, etc):

Expected: nothing should be unsubscribed.
Actual: some of the events processed before the error happened are unsubscribed.

[Related WPT test](https://github.com/web-platform-tests/wpt/blob/ef22a833e19a288616e478ebaffe6f346bc53826/webdriver/tests/bidi/session/unsubscribe/invalid.py#L75).

### PR DESCRIPTION
**Pull request by thiagowfx**
If an operation failed for some reason (not subscribed, browsing context not found, etc), do not unsubscribe partially.

Closes #482

### CODE REVIEW

**Comment by sadym-chromium** [comment]
Unit tests fail because of the wrong order of an event with different channels. It should be first subscribed - first received. Required for ChromeDriver Classic + BiDi

**Comment by thiagowfx** [comment]
Still - I didn't touch the `subscribe` method. I am trying to debug it but cannot see how changing `unsubscribe` affects those tests.

**Comment by thiagowfx** [comment]
The issue was `Promise.all`. Removed.

(An alternative could have been to just `await` upon calling `unsubscribe`.)

**Comment by thiagowfx** [comment]
This is now passing! :D

**Comment by thiagowfx** [comment]
Errr, sorry - this shouldn't have been merged before review, I made a mistake with the auto-merge (should not happen again!).

Please review this PR as usual and if there are comments I'll send a follow-up to address them.

File: src/bidiMapper/domains/events/SubscriptionManager.spec.ts [hunk 1]
```diff
@@ -52,6 +55,38 @@ describe('test SubscriptionManager', () => {
 
     subscriptionManager = new SubscriptionManager(browsingContextStorage);
   });
+
+  it('should re-subscribe global and specific event in proper order', () => {
```
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply src/bidiMapper/domains/events/SubscriptionManager.spec.ts:59]: Oh, it should be "subscribe twice"

File: src/bidiMapper/domains/events/SubscriptionManager.ts [hunk 2]
```diff
@@ -158,33 +195,57 @@ export class SubscriptionManager {
     eventMap.set(event, this.#subscriptionPriority++);
   }
 
+  /**
+   * Unsubscribes atomically from all events in the given contexts and channel.
+   */
+  unsubscribeAll(
```
Comment (sadym-chromium) [inline src/bidiMapper/domains/events/SubscriptionManager.ts:201]: Unit tests?

File: src/bidiMapper/domains/events/SubscriptionManager.ts [hunk 3]
```diff
@@ -158,33 +195,57 @@ export class SubscriptionManager {
...
+    events: Session.SubscribeParametersEvent[],
+    contextIds: (CommonDataTypes.BrowsingContext | null)[],
+    channel: string | null
+  ) {
+    // Assert all contexts are known.
+    for (const contextId of contextIds) {
+      if (contextId !== null) {
+        this.#browsingContextStorage.getKnownContext(contextId);
+      }
+    }
```
Comment (sadym-chromium) [inline src/bidiMapper/domains/events/SubscriptionManager.ts:211]: this check would logically fit better into `#checkUnsubscribe`

File: src/bidiMapper/domains/events/SubscriptionManager.ts [hunk 4]
```diff
@@ -158,33 +195,57 @@ export class SubscriptionManager {
...
+      ]
+    > = cartesianProduct(unrollEvents(events), contextIds);
+
+    // Assert all unsubscriptions are valid.
+    // If any of the unsubscriptions are invalid, do not unsubscribe from anything.
+    eventContextPairs
+      .map(([event, contextId]) =>
+        this.#checkUnsubscribe(event, contextId, channel)
+      )
+      .forEach((unsubscribe) => unsubscribe());
```
Comment (sadym-chromium) [inline src/bidiMapper/domains/events/SubscriptionManager.ts:226]: that is exactly what I meant, when we discussed it

Comment (thiagowfx) [reply src/bidiMapper/domains/events/SubscriptionManager.ts:226]: Glad to hear that!

File: src/bidiMapper/domains/events/SubscriptionManager.spec.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (thiagowfx) [reply src/bidiMapper/domains/events/SubscriptionManager.spec.ts:59]: Done, renamed

**Comment by thiagowfx** [comment]
The follow-up changes are here: https://github.com/GoogleChromeLabs/chromium-bidi/pull/511

File: src/bidiMapper/domains/events/SubscriptionManager.ts (hunk 3 above)
Comment (thiagowfx) [reply src/bidiMapper/domains/events/SubscriptionManager.ts:211]: It's more logical and it fits better, however the side effect is a small performance penalty. As of now:

- events: E1, E2
- contexts: C1, C2

Contexts are checked only twice: C1 and C2.

If we do what you propose, then it becomes:

[E1, C1], [E1, C2], [E2, C1], [E2, C2] --> call `checkUnsubscribe` for each of those --> Contexts are checked 4 times - 2 for C1 and 2 for C2.

If you don't care about the redundancy, then I'll go ahead.

Comment (thiagowfx) [reply src/bidiMapper/domains/events/SubscriptionManager.ts:211]: Actually, this change makes the unit test fail! https://github.com/GoogleChromeLabs/chromium-bidi/actions/runs/4325825970/jobs/7552477329

Comment (thiagowfx) [reply src/bidiMapper/domains/events/SubscriptionManager.ts:211]: Reverted this bit back to the original form.

File: src/bidiMapper/domains/events/SubscriptionManager.ts (hunk 2 above)
Comment (thiagowfx) [reply src/bidiMapper/domains/events/SubscriptionManager.ts:201]: Will do

Comment (thiagowfx) [reply src/bidiMapper/domains/events/SubscriptionManager.ts:201]: Done

**Pull request merged**